c++/src/kj/string.h

```diff
@@ -296,6 +306,105 @@ class String {
   Array<char> content;
 };
 
+// =======================================================================================
+// ConstString -- A NUL-terminated Array<const char> containing UTF-8 text.
+//
+// NUL bytes are allowed to appear before the end of the string.  The only requirement is that
+// a NUL byte appear immediately after the last byte of the content.  This terminator byte is not
+// counted in the string's size.
+//
+// A ConstString can be made by the explicit move from String constructor, or by providing a
+// StringPtr.
+
+class ConstString {
+public:
+  ConstString() = default;
+  inline ConstString(decltype(nullptr)): content(nullptr) {}
+  inline ConstString(const char* value, size_t size, const ArrayDisposer& disposer);
+  // Does not copy.  `size` does not include NUL terminator, but `value` must be NUL-terminated.
+  inline ConstString(kj::StringPtr ptr);
```

Comment:
I don't think we should so easily implicitly convert a StringPtr -- which may have limited lifetime -- to `ConstString` -- which seems like a type that is intended to own its backing buffer (like String does). This could easily lead to bugs.

I am comfortable with _specifically_ allowing string _literals_ to implicitly convert to `ConstString`. We could either do that by somehow making the `_kj` suffix magically able to convert to `ConstString` or, if that's too hard, by introducing a new `_kjc` suffix.

If we really need to give people a way to wrap StringPtrs that aren't literals, I would extend the `attach()` family of methods and have a `ConstString StringPtr::attach(...)`, where if you pass zero arguments, then it uses NullArrayDisposer (without allocating anything).

Reply:
> I am comfortable with *specifically* allowing string literals to implicitly convert to ConstString. We could either do that by somehow making the _kj suffix magically able to convert to ConstString

I'm not experienced enough in C++ to know what this magic might look like

> or, if that's too hard, by introducing a new _kjc suffix.

Couple of weird things with a specific suffix.

1. The operator can't be a `constexpr`, so it can't be used in other `constexpr`s
2. It can't be a `constexpr` because `kj::ConstString` isn't a literal type (because `kj::Array` isn't)

Reply:
How about this:

* Define a new subclass of StringPtr called LiteralStringPtr.
* Literals with the `_kj` suffix now evaluate to `LiteralStringPtr`. Indeed, this should be the only way to create a `LiteralStringPtr`.
* `LiteralStringPtr` defines an `operator ConstString()`, so it can also implicitly cast to `ConstString`.

Reply:
This messes up some existing code, like any code that expects a `Promise<StringPtr>` or possibly other generic types (`Promise<LiteralStringPtr>` won't automagically cast to a `Promise<StringPtr>`)

Reply:
Perhaps we could define another type called `LiteralStringConst` with the `_kjc` suffix?